Fix #6742: Only possible to build station next to competitor by using CTRL+click

Hemaolle opened the pull request:
This builds on @Hopman's original pull request (#6831) for by adding a fix suggested in that pull request for the case where there are two competitor stations nearby.

Closes #6831

Hopman committed:
Fix: Only possible to build station next to competitors by using CTRL+click

Fix by checking only for stations owned by the current company when
inspecting if there are multiple adjoining stations to the one being built.

When building next to 2 or more owned stations we don't know which
station should be extended. For other companies' stations that's not a
problem since our station won't merge with theirs anyway.

Calling to BuildStationPart should never have another company's station
as a parameter to attach to unless the client is malicious, so just returning
a generic error in that case.

LordAro commented:
I feel like the code would be cleaner if `GetOwnedStationAt` took the company ID as a parameter, rather than _current_company being hard coded in

Also, "there are no more than" (is->are)

LordAro commented:
As for commit messages, should probably be a Remove followed by a Fix, given those are the first words following the update :p

(Only need the issue number referenced once as well)

Hemaolle commented:
Can you clarify the comment about the commit messages? Are you referring to a specific commit message out of the 3 here? Do you mean that the commits themselves should be in a different order? Or are you referring to the message on the pull request perhaps?

Hemaolle commented:
I removed now the duplicated issue number from the PR message at least.

LordAro commented:
The commit messages themselves - "Update" is rarely used, and when it's immediately followed by another keyword (Fix, Remove) you can probably do a minor amount of rewording to use that instead. The order is fine

Hemaolle commented:
And only the first commit should have the issue number?

LordAro commented:
It doesn't really matter - the name of the PR itself should be enough

Hemaolle commented:
Addressed everything else besides "there are no more than" (is->are). For that I read: https://english.stackexchange.com/questions/35389/there-is-are-more-than-one-whats-the-difference and it seems to me based on that that my version would be correct.

LordAro reviewed (dismissed):
Ah, the English language, always with the edge cases (I say this as a native speaker)

I wonder about using the CompanyByte type instead of OwnerByte, but since they're identical and I don't know the different uses of them (there's probably some existing (mis)use of them anyway), I won't worry about it

michicc commented on src/station_cmd.cpp:
```diff
@@ -89,25 +89,25 @@ bool IsHangar(TileIndex t)
  */
 template <class T>
-CommandCost GetStationAround(TileArea ta, StationID closest_station, T **st)
+CommandCost GetStationAround(TileArea ta, StationID closest_station, OwnerByte owner, T **st)
```
Just a tiny nit-pick: The `...Byte` types are usually only used for save/load. In normal functions, `Owner` is preferred, even if just for aesthetics.

Hemaolle replied:
Changed this to CompanyID now, perhaps that's a bit more appropriate here. What do you think?

Hemaolle commented:
Changed the parameter new parameter to `GetStationAround` still from `OwnerByte` to `CompanyID`. How does that look?

frosch123 commented:
Are the first two commits needed? 
To me it looks like the third one fixes everything on its own.

Hemaolle commented:
@frosch123, you are probably right in that the third commit would fix the issue on its own. Nevertheless, I think the 2 other commits are good cleanup since `BuildStationPart` won't anymore be passed other companies' stations to attach to. For that reason it's kind of redundant to check in that method if the passed station is owned by the current company actually.

Hemaolle commented:
@LordAro would you mind reviewing this again. The previous review got discarded when I addressed michicc's comment.

nielsmh reviewed (changes requested):
Seems to work correctly, however please merge the two code changes and do a force-push, so there's just one "Fix bug" and one "Remove unused string" commit.

Hemaolle commented:
@nielsmh, done.

While at it, I removed the check for the `Station **st` parameter in `BuildStationPart()` being owned by the `_current_company` and replaced with an assert in the beginning as we won't be calling it anymore with other companies' stations to attach to.

frosch123 commented on src/station_cmd.cpp:
```diff
@@ -671,14 +671,12 @@ static void UpdateStationSignCoord(BaseStation *st)
  */
 static CommandCost BuildStationPart(Station **st, DoCommandFlag flags, bool reuse, TileArea area, StationNaming name_class)
 {
+	assert(*st == NULL || (*st)->owner == _current_company);
```
I think this assertion can be triggered by a malicious client to crash the server.
When distant-joining an existing station, the client sents the station id of the station to join.
At this point the id has been checked for existence, but not for ownership.

So, this must not be an "assert(...)" but a "if (..) return CMD_ERROR;"

Hemaolle replied:
Good thinking, looks like that could be the case indeed. Fixed it.

Hemaolle committed:
Remove: Unused string "STR_ERROR_TOO_CLOSE_TO_ANOTHER_STATION"
